File: credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long cc_number = get_long("Number: ");
    int digits[16];

    // Get length of cc number.
    int length_of_digits = 0;
    long num = cc_number;
    while (num != 0)
    {
        num /= 10;
        length_of_digits++;
    }

    // Add each digit of cc_number to array digits.
    for (int i = 0; cc_number; i++)
    {
        digits[i] = cc_number % 10;
        cc_number /= 10;
    }

    // Get the first and second digits of cc_number.
    int first_digit = digits[length_of_digits - 1];
    int second_digit = digits[length_of_digits - 2];

    // Calculate the sum of digits.
    int sum = 0;

    for (int i = 1; i < 16; i += 2)
    {
        if (digits[i] < 5)
        {
            sum = sum + (digits[i]) * 2;
        }
        else
        {
            sum = sum + (digits[i] * 2) % 10 + 1;
        }

    }
    for (int i = 0; i < 16; i += 2)
    {
        sum += digits[i];
    }

    // Check conditions for type and valid cc_number.
    if (sum % 10 == 0 && length_of_digits > 12)
    {
        if (first_digit == 3 && (second_digit == 4 || second_digit == 7))
        {
            printf("AMEX\n");
        }
        else if (first_digit == 4)
        {
            printf("VISA\n");
        }
        else if (first_digit == 5 && (second_digit == 1 || second_digit == 2 || second_digit == 3 || second_digit == 4
                                      || second_digit == 5))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }

    }
    else
    {
        printf("INVALID\n");
    }

}

File: pset6- hello, mario less, mario more, cash, credit, readability, dna/credit/credit.py
import string

from cs50 import get_string

cc_number = get_string("Number: ")

#ccRegex = re.compile(r'(\d{13}(\d{15,16})?)')
#mo = ccRegex.search(cc_number)

while (not cc_number.isdigit()):
    cc_number = get_string("Number: ")

# Multiply by 2 to every other digit and add digit sum to luhnSum
luhnSum = 0
for i in range(len(cc_number) - 2, -1, -2):
    productByTwo = int(cc_number[i]) * 2
    if productByTwo > 9:
        productStr = str(productByTwo)
        luhnSum = luhnSum + int(productStr[0]) + int(productStr[1])
    else:
        luhnSum += productByTwo

# Add remaining digits to luhnSum
for i in range(len(cc_number) - 1, -1, -2):
    luhnSum += int(cc_number[i])

# Check if AMEX, VISA, MASTERCARD or INVALID
if (len(cc_number) == 13 or len(cc_number) == 15 or len(cc_number) == 16) and luhnSum % 10 == 0:
    if len(cc_number) == 15 and (cc_number[:2] == '34' or cc_number[:2] == '37'):
        print("AMEX")
    elif len(cc_number) == 13 or len(cc_number) == 16 and cc_number[0] == '4':
        print("VISA")
    elif len(cc_number) == 16 and (int(cc_number[:2]) > 50 and int(cc_number[:2]) < 56):
        print("MASTERCARD")

else:
    print("INVALID")